### ISSUE TITLE
Deal with discord channel limits

### ISSUE THREAD

**Issue by richfromm** [richfromm/slack2discord#32]
While testing channel creation for https://github.com/richfromm/slack2discord/issues/24 (and somewhat for https://github.com/richfromm/slack2discord/issues/23 too), I came across yet another unexpected limit:
```
2023-02-08 11:51:26 ERROR    asyncio Task exception was never retrieved
future: <Task finished name='Task-5' coro=<MyClient.my_background_task() done, defined at /Users/rich/projects/slack2discord/poc/./background_task_asyncio.py:57> exception=HTTPException('400 Bad Request (error code: 50035): Invalid Form Body\nIn parent_id: Maximum number of channels in category reached (50)')>
Traceback (most recent call last):
  File "/Users/rich/projects/slack2discord/poc/./background_task_asyncio.py", line 69, in my_background_task
    await self.create_channels()
  File "/Users/rich/projects/slack2discord/poc/./background_task_asyncio.py", line 346, in create_channels
    await self.create_text_channel(guild, channel_name)
  File "/Users/rich/projects/slack2discord/poc/./background_task_asyncio.py", line 294, in create_text_channel
    channel = await guild.create_text_channel(channel_name, category=text_channels_category)
  File "/Users/rich/.virtualenvs/slack2discord/lib/python3.10/site-packages/discord/guild.py", line 1331, in create_text_channel
    data = await self._create_channel(
  File "/Users/rich/.virtualenvs/slack2discord/lib/python3.10/site-packages/discord/http.py", line 744, in request
    raise HTTPException(response, data)
discord.errors.HTTPException: 400 Bad Request (error code: 50035): Invalid Form Body
In parent_id: Maximum number of channels in category reached (50)
```

Googling a bit, I think the following limits exist and are relevant:
* Limit of 50 channels per category
* Limit of 500 channels per server

There's another limit that's worth noting here, but I don't think it's probably directly relevant:
* Limit of 50 categories per server (probably not relevant)

Regarding categories, the current behavior when creating any new channels is to look for a category with the name `'Text Channels'`. This is chosen because it is the default when creating a new Discord, it's where `#general` is located. But users (or maybe just admins?) can create new categories, and rename existing ones, so there's no guarantee that there will be such a category. If it is not found, we place any newly created channels in no category.

I verified that channels can be created with no category, and they are not subject to the 50 channel per category limit. Presumably the only limit is the 500 channels per server, but I admittedly haven't tested that yet.

In some ways the existing behavior is okay, but should perhaps consider a few modifications:
* Rather than just hardcode `'Text Channels'`, perhaps have that be the default, but allow overriding with a `--category` option.
* But what if the specified category can't be found? I think if it's a user inputted category, that's probably an error, not that there should be some fallback.
* Although another option would be to allow for the automatic creation of the category, just like we allow for the automatic creation of channels.
* But if we're going to allow for specification of category, do we also allow for specification of no category?

Sigh, already this is starting to get too complicated...

Let me move on to the more important points I wanted to make:
* If we are going to put newly created channels in a category, we should check in advance to make sure that we're not going hit the 50 channel per category limit.
* If we are going to hit the limit, then we should either fail, or create the channels with no category. I'm undecided about which, and/or to what extent this should be controllable.

So we could skip adding category features, keep it mostly the way it is, and just put new channels in `'Text Channels'` if it exists and there's room, or in no category if either that category doesn't exist, or if it does exist but doesn't have space for all of the new channels.

Regarding the 500 channel limit, we should also check that in advance, making sure that the the number of existing channels plus the number of channels that we are going to create doesn't put us over the limit. If it does, we have little choice but failing gracefully (ideally before changing **any** state in Discord), with a sufficiently instructive error message. Options include: the user can consolidate/delete channels from Slack, not import everything, import multiple Slack channels into a single Discord channel.

**Comment by richfromm** [comment]
Attn: @waocats @shmulvad

**Comment by waocats** [comment]
Might be something to prompt for depending on the user's choice in #23? If a category is full, you could ask if the user wants to create a new category (and put all of the transferred channels in there), or just place them outside of a category and let them do it within Discord.

**Comment by richfromm** [comment]
Yeah, that's what I was pondering, that any prompting could be done at the beginning as needed wrt channels, categories, or anything else that might come up that I haven't yet discovered.

**Comment by richfromm** [comment]
I've thought about it a bit, and I think I have an approach that's not overly complicated (although still a little more complicated than I'd like, judging by the length of this comment), with reasonably minimal cmd line options, that can simultaneously take care of https://github.com/richfromm/slack2discord/issues/23 https://github.com/richfromm/slack2discord/issues/24 and https://github.com/richfromm/slack2discord/issues/32.

* Determine, via the combination of the Slack export, the command line options, and the `--channel-file` if specified, the list of channels in Discord that we will need to post to.

* Look at which channels exist and which do not exist. If the number of existing channels, plus the number of channels that need to be created, is > 500, fail.

* For any channels that need to be created, make sure that they meet the naming criteria (alphanumeric plus dash plus underscore only, no multiple dashes in a row, total name length <= 100). If any fail validation, fail.

* For any channels that need to be created, they can or can not go in a category. The default behavior, if no command line option is specified, is to place the channels in a category named `'Text Channels'`. If that category does not exist, and if no command line option is specified, new channels will be created without a category. Additionally, if the number of channels in the `'Text Channels'` category plus the number of new channels is > 50, create the new channels not in any category.

* Add a new optional cmd line option `--category CATEGORY`. The user can use this to explicitly specify a category. If a category is explicitly specified, all newly created channels must go in that category. So if the number of existing channels in that category, plus the number of new channels, is > 50, fail, without an uncategorized fallback.

* If the category specified by the user does not exist, we will create it. However, if there are already 50 categories on the server, there's no room, so fail. Again, do not have an uncategorized fallback if the user has explicitly requested a category. (TODO: Determine what naming restrictions there are, if any, on categories.)

* Special-case the category name `None` (case insensitive?). If the user specifies `--category None`, that is explicitly stating to create any new channels without a category. (An alternative would be to have a separate option, `--no-category`, that would be mutually exclusive with `--category CATEGORY`.)

* Most of this discussion has been about newly created channels. For the existing channels, if a channel exists with the same name in multiple categories, we will pick, in order:
   * The one that matches the `--category CATEGORY` option if specified
   * The one that matches the `'Text Channels'` category, if no category specified
   * The one that is not in any category at all, if such a channel exists
   * Whatever is found first, if none of the other choices applies

* If we have to deal with the abomination of resolving channel naming collisions described above, we will inform the user of this as part of the point below.

* The default behavior is, before doing anything substantial (so before the call to `SlackParser.parse()`), present the user with a summary of the channel/category state. What channels are being posted to that already exist, what new channels are we creating, what categories these channels are in, etc. The user interactively can either accept or reject this. If accept, continue. If reject, fail.

* The user can bypass the above confirmation, with a new command line option. (Tentative possible naming suggestions: `--yes` or `--no-prompt`) This is the equivalent of accepting the prompt. Note that this implies that there's just a simple yes/no question, and not any kind of more complicated queries. It does not preclude a further expansion in the number of prompts, although my plan for now is to keep it as a single prompt. Even if the user automatically accepts this via a command line option, we should still output the information for later review.

I'm mildly worried that sufficiently documenting this behavior in a way that's not overly verbose (like this message) is going to be hard. I think the implementation may not actually be as complicated as this description makes it sound.

One more point I'm undecided is, for any category and channels that we need to create in Discord, when to do it. Up until now my strategy has been to delay making **any** changes to Discord until as late as possible. This entire process is nowhere near atomic, but by getting all state from Slack before putting any new state in Discord, it at least moved it a step in that direction. We should definitely check things and prompt the user right away (before we even parse Slack). The question is do we then immediately create any channels (and possibly category) in Discord if this is needed, or defer this (like we do now) until we're about to post to Discord.

I can see arguments either way. Given the fact that the script isn't designed that well in terms of being re-run (although somewhat better now with https://github.com/richfromm/slack2discord/issues/26), that argues somewhat for creating early and failing early if something goes wrong. And there may be more limits that I'm not even aware of.

Wanting to be somewhat more atomic argues for creating as late as possible. The async nature of the `DiscordClient` class also complicates things a bit. Right now there's a a single background task (`post_messages()`) that takes care of all of the work that the client needs to do, when `DiscordClient.run()` is called. Doing otherwise would require some refactoring. Getting the client working properly took a bit of trial and error, and I'm not sure if I want to revisit that right now.

**Comment by richfromm** [comment]
Sigh, this is going to be harder than I thought.

I did some experimenting with the discord client, and again got frustrated getting it to behave the way I want, so I was about to say it's not worth it, and I should just defer category/channel creation until later.

But that's not a good enough answer. Even just to do a bunch of this checking (e.g. get the list of channels), I need a functioning discord client earlier.

Basically, what I want is something like:
```
client = DiscordClient(params)
client.do_some_stuff()
# ... other non-client code
client.do_some_other_stuff()
```

You might think that ought to be easy, but it's not.

Part of the problem is that the Discord API isn't really designed for what I'm doing. It's meant for writing bots. You set things up, the client itself is a long running process, a websocket is created and you can get callbacks when various [events](https://discordpy.readthedocs.io/en/latest/api.html#event-reference) happen, like a user sends a message.

But what I want to do is to just be able to synchronously do various actions, like query state, or post messages, where I initiate the action.

I'm doing this by abusing a hook during [setup](https://discordpy.readthedocs.io/en/latest/api.html?highlight=setup_hook#discord.Client.setup_hook), I create a single task to do what I want and add it to the event loop. That executes when I call `run()`. But it blocks, and doesn't return until I call `close()`. And at that point it's too late to do anything more with the client. Basically, I'm doing:
```
client = DiscordClient(params)
client.run()
```

Everything using the client all happens in one chunk.

Perhaps I need to be making lower level API calls, managing the event loop somewhat manually. [This reply](https://github.com/Rapptz/discord.py/issues/516#issuecomment-287511041) might have some helpful info, as well as the differences between [`run()`](https://discordpy.readthedocs.io/en/latest/api.html?highlight=setup_hook#discord.Client.run) and [`start()`](https://discordpy.readthedocs.io/en/latest/api.html?highlight=setup_hook#discord.Client.start). Or perhaps I need to read more about how asyncio and the event loop work -- I admittedly do not have the best understanding of it.

I can also think of some hacky options, like:
* create a different client instance every time I want to do something, and a parameter that it takes is the function that I want to pass to `self.loop.create_task()` in `DiscordClient.setup_hook()`, but that seems somewhat wasteful.
* make the client be kind of the master of all of slack2discord, and everything happens from that starting point. that seems like a horrible software architecture

I'm happy to take any advice about working with the Discord API in a non-bot related context. I might also try asking the discord.py folks for [help](https://discordpy.readthedocs.io/en/latest/#getting-help).

**Comment by waocats** [comment]
Worst case, I suppose you could use the web API yourself with no wrapper (just using `requests`), but it's definitely not the most amazing option (and is somewhat reinventing the wheel). Asking the folks at `discord.py` might be the best way to go, here--but I have the feeling they'll tell you that you'd need your client to be initialized at the start and be globally accessible throughout the program (either that or you'd just need to create a new client every time). As you said, `discord.py` probably wasn't really made with this use case in mind.

**Comment by richfromm** [comment]
Yeah, I thought about just doing direct HTTP API calls. Which wouldn't be **that** hard, but I suspect there would be a bit too much reinventing the wheel, esp. with things like error handling.

I did some reading and experimenting today, and I'm coming to the conclusion that what I ideally want to do just isn't possible. `discord.py` is very much an async library, with a different goal in mind than what I'm doing, and crossing the sync/async boundary is hard. I can't find any way to get synchronous control back from the client without the event loop having stopped. Which effectively means that different synchronous client invocations need to have different client instantiations.

[This response](https://github.com/Rapptz/discord.py/discussions/8227#discussioncomment-3163363) further makes the point, going so far as to say:

> I personally can't think of a context in which you would need to use the Discord API without a Discord bot present

(The tl;dr of the response is it basically says either wrap your each function call within the library with sync calls or use the HTTP API instead of `discord.py`)

I tried some of the suggested workaround with `discord.commands.ext.Bot`, but it seems way too hacky, and there's some functionality I need that I wasn't able to get to work after a bit of effort, and I didn't feel like trying any further.

I think I can get something to work that's closer to the model I have now. I've already made it clear that my `DiscordClient` isn't trying to be a generic client, but is specifically written with the importing to Discord from a Slack export in mind. It won't be that bad to separate the functionality I want into two parts. The first part gets run early and just reads state from Discord, and the second part gets run later and changes state in Discord (creating channels and/or a category if needed, and posting messages). Each requires a separate instantiation of a client object. Which is mildly inefficient, but this is something just run once, and it's only two instantiations, so it's not really that big of a deal, it's not that important to optimize for performance. (Which was somewhat the argument I used against parallelizing the file downloads, I'd rather keep simplicity over raw performance.) Putting the two different functionalities in a small class hierarchy, and/or using class variables to pass state from the first to the second case, might make for a slightly cleaner implementation.

[This response](https://stackoverflow.com/a/52569809/9797192) to a similar question suggests [janus](https://github.com/aio-libs/janus) as a means of integrating sync and async code, but I think my needs are limited enough (just the two cases I outlined above) that I don't need to introduce some general solution that I suspect would ultimately be more complicated.
